Focus item click conversion rate

patrykgruszka opened the pull request:
<!-- ## Which branch should I use for my PR?

Assuming that:

a = current major release
b = current minor release
c = future major release

* a.x for any features and enhancements (e.g. 5.x)
* a.b for any bug fixes (e.g. 4.4, 5.1)
* c.x for any features, enhancements or bug fixes with backward compatibility breaking changes (e.g. 5.x) -->

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [ N ]
| New feature/enhancement? (use the a.x branch)      | [ Y ]
| Deprecations?                          | [ N ]
| BC breaks? (use the c.x branch)        | [ N ]
| Automated tests included?              | [ Y ] <!-- All PRs must maintain or improve code coverage -->
| Related user documentation PR URL      | mautic/mautic-documentation#... <!-- required for new features -->
| Related developer documentation PR URL | mautic/developer-documentation#... <!-- required for developer-facing changes -->
| Issue(s) addressed                     | Fixes #... <!-- prefix each issue number with "Fixes #", no need to create an issue if none exists, explain below instead -->

<!--
Additionally (see https://contribute.mautic.org/contributing-to-mautic/developer/code/pull-requests#work-on-your-pull-request):
 - Always add tests and ensure they pass.
 - Bug fixes must be submitted against the lowest maintained branch where they apply
   (lowest branches are regularly merged to upper ones so they get the fixes too.)
 - Features and deprecations must be submitted against the "4.x" branch.
-->

#### Description:
This PR adds conversion rate stat to links in focus view and total views count in focus details.
![image](https://user-images.githubusercontent.com/8580942/221558963-c257dd83-6ae0-4412-aac1-416917df4709.png)

<!--
Please write a short README for your feature/bugfix. This will help people understand your PR and what it aims to do. If you are fixing a bug and if there is no linked issue already, please provide steps to reproduce the issue here.
-->

#### Steps to test this PR:

<!--
This part is really important. If you want your PR to be merged, take the time to write very clear, annotated and step by step test instructions. Do not assume any previous knowledge - testers may not be developers.
-->
1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Add focus item with **Emphasize a link** type and add some links to the content
3. Copy the focus installation script
4. Create new landing page and paste the focus installation script in the **Script inside head** field
5. Open the landing page using the link from **Page URL** and use private tab
6. Click the link from LP and generate some more statistics
7. Check if the statistics are correct

<!--
If you have any deprecations, list them here along with the new alternative.
If you have any backwards compatibility breaks, list them here.
-->

patrykgruszka committed:
feat: [DPMMA-1742] focus item conversion rate

patrykgruszka committed:
feat: [DPMMA-1742] FocusModel functional tests

patrykgruszka committed:
Merge branch '5.x' into DPMMA-1742_focus-item-stats

patrykgruszka committed:
Merge branch '5.x' into DPMMA-1742_focus-item-stats

patrykgruszka committed:
fix: [DPMMA-1742] migrate to twig

kuzmany reviewed (dismissed):
Good to me 👍

escopecz reviewed (changes requested):
From the past experience with similar stats on other detail pages (email, campaign, ...) data like this will cause troubles:
1. The count query is very time consuming when the table size goes to millions and more. We are optimizing count queries that take minutes for our clients. Meaning that if we merge this I see a priority ticket following right after saying that some focus item pages are loading way to slow or timing out. How to solve this? Caching and/or loading it asynchronously.
2. The number of hits can be higher than number of focus stats. So the conversion rate will be confusing to users. Imagine that you view the focus item once and click the link twice. Which will result in more support tickets. How about to use `unique_hits` instead of `hits`?
3. It will be also confusing for users whether the conversion rate shows numbers for the selected time period or total.
4. Why the 4 decimal number rounding? I think that 1 would be enough. 2 max.

patrykgruszka committed:
Merge branch '5.x' into DPMMA-1742_focus-item-stats

patrykgruszka committed:
fix: [DPMMA-1742] ajax endpoint focus:getViewsCountAction

patrykgruszka committed:
fix: [DPMMA-1742] async load conversion rate in focus items

patrykgruszka committed:
fix: [DPMMA-1742] use cache for focus item views count ajax action

patrykgruszka committed:
fix: [DPMMA-1742] fix phpstan

patrykgruszka commented:
@escopecz, You are right, I have fixed these problems:
1. Added caching and loading data asynchronously
2. Changed to `unique_hits`
3. Click Counts table refers to the total. I don't think the Conversion Rate makes a difference here.
4. Changed to 2 decimal places

Syrgak-Alan commented:
I made all step by step, and nothing happened. Focus items do not work correctly, sometimes they are not displayed at all on the main page of the landing page 
![image](https://user-images.githubusercontent.com/96595532/236477381-b4abbc48-7bed-458e-9383-31565111e114.png)

patrykgruszka commented:
@Syrgak-Alan this PR does not change the behaviour of the landing page functionality. Most likely there is a problem with your landing page/tracking or a bug introduced in version 5. 

I can help you test this feature, but I need more information about your problem. Have you tried opening the LP in a private browser session? Tracking will not work properly if you are logged into Mautic at the same time.

jos0405 commented:
Hmm. I can't save the focus item, getting:
[2023-05-12 07:09:01] mautic.CRITICAL: Uncaught PHP Exception Twig\Error\LoaderError: "Unable to find template "email.html.twig" (looked into: /var/www/html/vendor/knplabs/knp-menu/src/Knp/Menu/Resources/views, /var/www/html/vendor/symfony/twig-bridge/Resources/views/Form)." at /var/www/html/app/bundles/FormBundle/Resources/views/Builder/form.html.twig line 75 {"exception":"[object] (Twig\\Error\\LoaderError(code: 0): Unable to find template \"email.html.twig\" (looked into: /var/www/html/vendor/knplabs/knp-menu/src/Knp/Menu/Resources/views, /var/www/html/vendor/symfony/twig-bridge/Resources/views/Form). at /var/www/html/app/bundles/FormBundle/Resources/views/Builder/form.html.twig:75)
[stacktrace]
#0 /var/www/html/vendor/symfony/twig-bundle/Loader/FilesystemLoader.php(82): Twig\\Loader\\FilesystemLoader->findTemplate('email.html.twig')
#1 /var/www/html/vendor/twig/twig/src/Loader/FilesystemLoader.php(131): Symfony\\Bundle\\TwigBundle\\Loader\\FilesystemLoader->findTemplate('email.html.twig')
#2 /var/www/html/vendor/twig/twig/src/Environment.php(264): Twig\\Loader\\FilesystemLoader->getCacheKey('email.html.twig')
#3 /var/www/html/vendor/twig/twig/src/Environment.php(312): Twig\\Environment->getTemplateClass('email.html.twig')
#4 /var/www/html/vendor/twig/twig/src/Environment.php(436): Twig\\Environment->load('email.html.twig')
#5 /var/www/html/vendor/twig/twig/src/Extension/CoreExtension.php(1340): Twig\\Environment->resolveTemplate('email.html.twig')
#6 /var/www/html/vendor/twig/twig/src/Environment.php(361) : eval()'d code(174): twig_include(Object(Twig\\Environment), Array, 'email.html.twig', Array)
#7 /var/www/html/vendor/twig/twig/src/Template.php(394): __TwigTemplate_a8c367275c8badda9c2816dabba683209e077e580c22a22ca5a5d3c0ed672329->doDisplay(Array, Array)
#8 /var/www/html/vendor/twig/twig/src/Template.php(367): Twig\\Template->displayWithErrorHandling(Array, Array)
#9 /var/www/html/vendor/twig/twig/src/Template.php(379): Twig\\Template->display(Array)
#10 /var/www/html/vendor/twig/twig/src/TemplateWrapper.php(40): Twig\\Template->render(Array, Array)
#11 /var/www/html/vendor/twig/twig/src/Extension/CoreExtension.php(1347): Twig\\TemplateWrapper->render(Array)
#12 /var/www/html/vendor/twig/twig/src/Environment.php(361) : eval()'d code(197): twig_include(Object(Twig\\Environment), Array, '@MauticForm/Bui...', Array, false)
#13 /var/www/html/vendor/twig/twig/src/Template.php(394): __TwigTemplate_f2df425b7fbcbd65d8ed5276513f6f565c4ce4293eb851becaef62fbec456419->doDisplay(Array, Array)
#14 /var/www/html/vendor/twig/twig/src/Template.php(367): Twig\\Template->displayWithErrorHandling(Array, Array)
#15 /var/www/html/vendor/twig/twig/src/Template.php(379): Twig\\Template->display(Array)
#16 /var/www/html/vendor/twig/twig/src/TemplateWrapper.php(40): Twig\\Template->render(Array, Array)
#17 /var/www/html/vendor/twig/twig/src/Environment.php(280): Twig\\TemplateWrapper->render(Array)
#18 /var/www/html/plugins/MauticFocusBundle/Model/FocusModel.php(290): Twig\\Environment->render('@MauticFocus/Bu...', Array)
#19 /var/www/html/plugins/MauticFocusBundle/Controller/AjaxController.php(48): MauticPlugin\\MauticFocusBundle\\Model\\FocusModel->getContent(Array, true)
#20 /var/www/html/app/bundles/CoreBundle/Controller/AjaxController.php(128): MauticPlugin\\MauticFocusBundle\\Controller\\AjaxController->generatePreviewAction(Object(Symfony\\Component\\HttpFoundation\\Request), 'focus')
#21 /var/www/html/vendor/symfony/http-kernel/HttpKernel.php(169): Mautic\\CoreBundle\\Controller\\AjaxController->executeAjaxAction('generatePreview', Object(Symfony\\Component\\HttpFoundation\\Request), 'focus')
#22 /var/www/html/vendor/symfony/http-kernel/HttpKernel.php(81): Symfony\\Component\\HttpKernel\\HttpKernel->handleRaw(Object(Symfony\\Component\\HttpFoundation\\Request), 2)
#23 /var/www/html/vendor/symfony/framework-bundle/Controller/ControllerTrait.php(96): Symfony\\Component\\HttpKernel\\HttpKernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 2)
#24 /var/www/html/app/bundles/CoreBundle/Controller/AjaxController.php(106): Symfony\\Bundle\\FrameworkBundle\\Controller\\AbstractController->forward('MauticPlugin\\\\Ma...', Array)
#25 /var/www/html/vendor/symfony/http-kernel/HttpKernel.php(169): Mautic\\CoreBundle\\Controller\\AjaxController->delegateAjaxAction()
#26 /var/www/html/vendor/symfony/http-kernel/HttpKernel.php(81): Symfony\\Component\\HttpKernel\\HttpKernel->handleRaw(Object(Symfony\\Component\\HttpFoundation\\Request), 1)
#27 /var/www/html/vendor/symfony/http-kernel/Kernel.php(201): Symfony\\Component\\HttpKernel\\HttpKernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#28 /var/www/html/app/AppKernel.php(105): Symfony\\Component\\HttpKernel\\Kernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#29 /var/www/html/app/middlewares/CORSMiddleware.php(82): AppKernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#30 /var/www/html/app/middlewares/HSTSMiddleware.php(35): Mautic\\Middleware\\CORSMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#31 /var/www/html/app/middlewares/CatchExceptionMiddleware.php(34): Mautic\\Middleware\\HSTSMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#32 /var/www/html/app/middlewares/Dev/IpRestrictMiddleware.php(52): Mautic\\Middleware\\CatchExceptionMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#33 /var/www/html/app/middlewares/VersionCheckMiddleware.php(58): Mautic\\Middleware\\Dev\\IpRestrictMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#34 /var/www/html/app/middlewares/TrustMiddleware.php(42): Mautic\\Middleware\\VersionCheckMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#35 /var/www/html/vendor/stack/builder/src/Stack/StackedHttpKernel.php(23): Mautic\\Middleware\\TrustMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#36 /var/www/html/vendor/stack/run/src/Stack/run.php(13): Stack\\StackedHttpKernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request))
#37 /var/www/html/index_dev.php(25): Stack\
un(Object(Stack\\StackedHttpKernel))
#38 {main}
"} {"hostname":"mautic-web","pid":3670}
[2023-05-12 07:09:10] mautic.CRITICAL: Uncaught PHP Exception Twig\Error\LoaderError: "Unable to find template "email.html.twig" (looked into: /var/www/html/vendor/knplabs/knp-menu/src/Knp/Menu/Resources/views, /var/www/html/vendor/symfony/twig-bridge/Resources/views/Form)." at /var/www/html/app/bundles/FormBundle/Resources/views/Builder/form.html.twig line 75 {"exception":"[object] (Twig\\Error\\LoaderError(code: 0): Unable to find template \"email.html.twig\" (looked into: /var/www/html/vendor/knplabs/knp-menu/src/Knp/Menu/Resources/views, /var/www/html/vendor/symfony/twig-bridge/Resources/views/Form). at /var/www/html/app/bundles/FormBundle/Resources/views/Builder/form.html.twig:75)
[stacktrace]
#0 /var/www/html/vendor/symfony/twig-bundle/Loader/FilesystemLoader.php(82): Twig\\Loader\\FilesystemLoader->findTemplate('email.html.twig')
#1 /var/www/html/vendor/twig/twig/src/Loader/FilesystemLoader.php(131): Symfony\\Bundle\\TwigBundle\\Loader\\FilesystemLoader->findTemplate('email.html.twig')
#2 /var/www/html/vendor/twig/twig/src/Environment.php(264): Twig\\Loader\\FilesystemLoader->getCacheKey('email.html.twig')
#3 /var/www/html/vendor/twig/twig/src/Environment.php(312): Twig\\Environment->getTemplateClass('email.html.twig')
#4 /var/www/html/vendor/twig/twig/src/Environment.php(436): Twig\\Environment->load('email.html.twig')
#5 /var/www/html/vendor/twig/twig/src/Extension/CoreExtension.php(1340): Twig\\Environment->resolveTemplate('email.html.twig')
#6 /var/www/html/vendor/twig/twig/src/Environment.php(361) : eval()'d code(174): twig_include(Object(Twig\\Environment), Array, 'email.html.twig', Array)
#7 /var/www/html/vendor/twig/twig/src/Template.php(394): __TwigTemplate_a8c367275c8badda9c2816dabba683209e077e580c22a22ca5a5d3c0ed672329->doDisplay(Array, Array)
#8 /var/www/html/vendor/twig/twig/src/Template.php(367): Twig\\Template->displayWithErrorHandling(Array, Array)
#9 /var/www/html/vendor/twig/twig/src/Template.php(379): Twig\\Template->display(Array)
#10 /var/www/html/vendor/twig/twig/src/TemplateWrapper.php(40): Twig\\Template->render(Array, Array)
#11 /var/www/html/vendor/twig/twig/src/Extension/CoreExtension.php(1347): Twig\\TemplateWrapper->render(Array)
#12 /var/www/html/vendor/twig/twig/src/Environment.php(361) : eval()'d code(197): twig_include(Object(Twig\\Environment), Array, '@MauticForm/Bui...', Array, false)
#13 /var/www/html/vendor/twig/twig/src/Template.php(394): __TwigTemplate_f2df425b7fbcbd65d8ed5276513f6f565c4ce4293eb851becaef62fbec456419->doDisplay(Array, Array)
#14 /var/www/html/vendor/twig/twig/src/Template.php(367): Twig\\Template->displayWithErrorHandling(Array, Array)
#15 /var/www/html/vendor/twig/twig/src/Template.php(379): Twig\\Template->display(Array)
#16 /var/www/html/vendor/twig/twig/src/TemplateWrapper.php(40): Twig\\Template->render(Array, Array)
#17 /var/www/html/vendor/twig/twig/src/Environment.php(280): Twig\\TemplateWrapper->render(Array)
#18 /var/www/html/plugins/MauticFocusBundle/Model/FocusModel.php(290): Twig\\Environment->render('@MauticFocus/Bu...', Array)
#19 /var/www/html/plugins/MauticFocusBundle/Model/FocusModel.php(207): MauticPlugin\\MauticFocusBundle\\Model\\FocusModel->getContent(Array, false, '')
#20 /var/www/html/plugins/MauticFocusBundle/Model/FocusModel.php(166): MauticPlugin\\MauticFocusBundle\\Model\\FocusModel->generateJavascript(Object(MauticPlugin\\MauticFocusBundle\\Entity\\Focus))
#21 /var/www/html/app/bundles/CoreBundle/Controller/AbstractStandardFormController.php(992): MauticPlugin\\MauticFocusBundle\\Model\\FocusModel->saveEntity(Object(MauticPlugin\\MauticFocusBundle\\Entity\\Focus))
#22 /var/www/html/plugins/MauticFocusBundle/Controller/FocusController.php(49): Mautic\\CoreBundle\\Controller\\AbstractStandardFormController->newStandard()
#23 /var/www/html/app/bundles/CoreBundle/Controller/CommonController.php(458): MauticPlugin\\MauticFocusBundle\\Controller\\FocusController->newAction(0, '')
#24 /var/www/html/vendor/symfony/http-kernel/HttpKernel.php(169): Mautic\\CoreBundle\\Controller\\CommonController->executeAction('new', 0, 0, '')
#25 /var/www/html/vendor/symfony/http-kernel/HttpKernel.php(81): Symfony\\Component\\HttpKernel\\HttpKernel->handleRaw(Object(Symfony\\Component\\HttpFoundation\\Request), 1)
#26 /var/www/html/vendor/symfony/http-kernel/Kernel.php(201): Symfony\\Component\\HttpKernel\\HttpKernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#27 /var/www/html/app/AppKernel.php(105): Symfony\\Component\\HttpKernel\\Kernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#28 /var/www/html/app/middlewares/CORSMiddleware.php(82): AppKernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#29 /var/www/html/app/middlewares/HSTSMiddleware.php(35): Mautic\\Middleware\\CORSMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#30 /var/www/html/app/middlewares/CatchExceptionMiddleware.php(34): Mautic\\Middleware\\HSTSMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#31 /var/www/html/app/middlewares/Dev/IpRestrictMiddleware.php(52): Mautic\\Middleware\\CatchExceptionMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#32 /var/www/html/app/middlewares/VersionCheckMiddleware.php(58): Mautic\\Middleware\\Dev\\IpRestrictMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#33 /var/www/html/app/middlewares/TrustMiddleware.php(42): Mautic\\Middleware\\VersionCheckMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#34 /var/www/html/vendor/stack/builder/src/Stack/StackedHttpKernel.php(23): Mautic\\Middleware\\TrustMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#35 /var/www/html/vendor/stack/run/src/Stack/run.php(13): Stack\\StackedHttpKernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request))
#36 /var/www/html/index_dev.php(25): Stack\
un(Object(Stack\\StackedHttpKernel))
#37 {main}
"} {"hostname":"mautic-web","pid":3658}

patrykgruszka committed:
Merge branch '5.x' into DPMMA-1742_focus-item-stats

patrykgruszka commented:
@jos0405 I've just tested this on Gitpod and haven't got this problem. I tried to update my branch locally to 5.x, but got some problems with loading with JS after recent changes. I'll update this PR once it's fixed and let you know when it's ready to test again. Thanks for your comment!

escopecz commented:
@patrykgruszka please check if it wasn't https://github.com/mautic/mautic/pull/12337 or let me know what is the issue you are experiencing

patrykgruszka committed:
Merge branch '5.x' into DPMMA-1742_focus-item-stats

patrykgruszka commented:
@escopecz I just tried the latest 5.x on Gitpod and this problem still exists - looks like a JS loading issue. Am I the only one having this problem? 
![image](https://github.com/mautic/mautic/assets/8580942/37794b59-6d82-4009-86e0-7eff37a13aaa)

EDIT: I've just created a new issue #12349 as this issue is not related to this PR.

patrykgruszka committed:
Merge branch '5.x' into DPMMA-1742_focus-item-stats

patrykgruszka committed:
fix: [DPMMA-1742] cache provider dependency injection

patrykgruszka committed:
fix: [DPMMA-1742] no views fix

patrykgruszka committed:
fix: [DPMMA-1742] twig template fix

patrykgruszka committed:
fix: [DPMMA-1742] twig template fix

patrykgruszka commented:
I've found some problems with the Focus Item twig template, it should be fixed now. Code is updated to latest changes from the 5.x branch. Tested on Gitpod and works fine for me - I've updated the instructions in this PR description. Ready to retest.

Syrgak-Alan commented:
@patrykgruszka  
![Bildschirmfoto 2023-05-16 um 15 50 43](https://github.com/mautic/mautic/assets/96595532/e984f90f-34ed-4f0c-b862-cd3d5b667aba)
I tried again and everything works great! But in the description for the test of this PR, it was written to insert the Focus Installation script into the head block. I tried to do it like this, but the changes just don't get saved inside the LP builder, so I just put in the landing page as {focus=1}

patrykgruszka commented:
Thanks @Syrgak-Alan for your feedback. I see that there are still some problems. The conversion rate should not be higher than 100%. I'll look into it.

Syrgak-Alan commented:
@patrykgruszka 
it's 200%, probably because I clicked on the link 3 times and each of the clicks counted as a separate conversion

patrykgruszka committed:
fix: [DPMMA-1742] total conversion rate fix

patrykgruszka committed:
fix: [DPMMA-1742] phpstan

patrykgruszka commented:
@Syrgak-Alan yes, but a contact should only be counted once for the total conversion rate. I've fixed this and added a tooltip showing how it's calculated:
![image](https://github.com/mautic/mautic/assets/8580942/df8059d9-f818-446a-83d4-93a9c9961476)

patrykgruszka committed:
fix: [DPMMA-1742] change total conversion rate tooltip

patrykgruszka committed:
Merge branch '5.x' into DPMMA-1742_focus-item-stats

patrykgruszka committed:
Merge branch '5.x' into DPMMA-1742_focus-item-stats

magdalenaleonow reviewed (changes requested):
Tested, found some issues:

1. In case where 3 contacts viewed the page with Focus item, one of them twice:

![image](https://github.com/mautic/mautic/assets/130159751/bea085d8-c307-4781-8a40-e3bbffa1912f)
Two of those contacts clicked on links:
![image](https://github.com/mautic/mautic/assets/130159751/496cc25f-8898-4c38-8e89-05e068c5ff6e)
The total Conversion Rate should be 2/3=66.6%, since we should take unique views (3, not 4 in this case)

2. In the tooltip for Total conversion rate the description should be written the other way around: 
Number of contacts that clicked on at least one link/Total views

3. Views in the table are not refreshed immediately with the rest of the data:

- Cached data timeout in System settings is bigger than 0
- One contact clicks on all of the links
- Wait for the data to be refreshed (views too)
- Second contact clicks on all of the links 
- Views is still 1, the rest of the data is updated immediately.
- While we wait for views to refresh Conversion rate for the links is 200%:
![image](https://github.com/mautic/mautic/assets/130159751/9c4604a2-1ae5-4c11-aeae-ec15c3a5c845)

patrykgruszka committed:
fix: [DPMMA-1742] conversion rate fixes

patrykgruszka committed:
fix: [DPMMA-1742] add focus item trackables cache

patrykgruszka committed:
fix: [DPMMA-1742] focus item view caching

patrykgruszka commented:
@magdalenaleonow nice shot with this tests. I've fixed the issue with calculating the rates.

About this cache issue, I think we should either cache everything or nothing to keep the data consistent within the view. I think we should go for the first option. I have implemented this and PR is ready to test.

magdalenaleonow reviewed (approved):
@patrykgruszka retests done, everything works
![image](https://github.com/mautic/mautic/assets/130159751/037c092f-6306-4409-bfaf-47426b4f755b)

LeonOltmanns reviewed (commented):
Just tested this feature using GitPod and works great, but the Preview for Focus Items in the FI-Builder was broken and following error got logged:

`[2023-07-22 09:11:18] mautic.CRITICAL: Uncaught PHP Exception InvalidArgumentException: "The controller for URI "/s/ajax" is not callable: Method "generatePreviewAction" on class "MauticPlugin\MauticFocusBundle\Controller\AjaxController" should be public and non-abstract." at /var/www/html/vendor/symfony/http-kernel/Controller/ControllerResolver.php line 88 {"exception":"[object] (InvalidArgumentException(code: 0): The controller for URI \"/s/ajax\" is not callable: Method \"generatePreviewAction\" on class \"MauticPlugin\\MauticFocusBundle\\Controller\\AjaxController\" should be public and non-abstract. at /var/www/html/vendor/symfony/http-kernel/Controller/ControllerResolver.php:88, InvalidArgumentException(code: 0): Method \"generatePreviewAction\" on class \"MauticPlugin\\MauticFocusBundle\\Controller\\AjaxController\" should be public and non-abstract. at /var/www/html/vendor/symfony/http-kernel/Controller/ControllerResolver.php:134)"} {"hostname":"mautic-web","pid":2078}
`

Please check wether this is just an issue with the GitPod testing instance or something of more importance.

RCheesley commented:
I am seeing the same error as @oltmanns-leuchtfeuer above when I save a focus item with this PR applied. It does actually let me create the focus item though, despite seeing an error.

When I followed the steps I am seeing the clicks I expect to see but I'm not sure why there's two links when testing with Gitpod (which I see with a few other screenshots above) - maybe a quirk of Gitpod?

![screenshot-8080-patrykgruszka-mautic-zdzzvxcui35 ws-eu102 gitpod io-2023 07 22-22_08_01](https://github.com/mautic/mautic/assets/2930593/2737f6db-c52b-4bb5-969c-d022d07b19f9)

RCheesley commented:
@oltmanns-leuchtfeuer the bug is present in the 5.x branch so I raised it here: https://github.com/mautic/mautic/issues/12577

escopecz committed:
Merge branch '5.x' into DPMMA-1742_focus-item-stats

escopecz commented:
@oltmanns-leuchtfeuer @RCheesley I fixed this error at https://github.com/escopecz/mautic/commit/9be9da4669792578541f7f18b2ab39d331e648b0. I merged the 5.x branch so it should be ready to retest now.

LeonOltmanns commented:
Hi @escopecz, I just re-tested it in GitPod and cannot access the Focus Item Menu-Point anymore. In the top right corner it says: "Your request could not be processed, please try again".

This was logged:
`[2023-07-24 11:44:09] mautic.CRITICAL: Uncaught PHP Exception TypeError: "Mautic\CoreBundle\Controller\AbstractStandardFormController::__construct(): Argument #1 ($formFactory) must be of type Symfony\Component\Form\FormFactoryInterface, Mautic\CoreBundle\Security\Permissions\CorePermissions given, called in /var/www/html/plugins/MauticFocusBundle/Controller/FocusController.php on line 38" at /var/www/html/app/bundles/CoreBundle/Controller/AbstractStandardFormController.php line 32 {"exception":"[object] (TypeError(code: 0): Mautic\\CoreBundle\\Controller\\AbstractStandardFormController::__construct(): Argument #1 ($formFactory) must be of type Symfony\\Component\\Form\\FormFactoryInterface, Mautic\\CoreBundle\\Security\\Permissions\\CorePermissions given, called in /var/www/html/plugins/MauticFocusBundle/Controller/FocusController.php on line 38 at /var/www/html/app/bundles/CoreBundle/Controller/AbstractStandardFormController.php:32)"} {"hostname":"mautic-web","pid":2095}`

escopecz committed:
Fixing controller dependencies

escopecz commented:
@patrykgruszka I updated the composer dependencies that have changed in M5-alpha but there are more code to update. Please check the failing CI.

patrykgruszka committed:
fix: [DPMMA-1742] fix phpstan

patrykgruszka committed:
Merge branch '5.x' into DPMMA-1742_focus-item-stats

patrykgruszka commented:
@escopecz fixed the failing phpstan, but I need to have a closer look at the problem described by @RCheesley. Links should not be duplicated.

patrykgruszka committed:
fix: [DPMMA-1742] fix double trackables

patrykgruszka commented:
@escopecz I have resolved two problems that are also on the 5.x branch:
1. When you create a focus item with **Emphasize a link** the link is always generated in preview mode. You can create a focus item, use it on LP, but the link will be always '#'. This is caused by a new twig template, fixed in https://github.com/mautic/mautic/pull/11945/commits/22d42f3b8ea1e2e245050c53a77ffbd5307dc85f
2. If you solve problem 1, then you will run into the double trackable problem that has been there since version 4.x. This is because the trackable is created first in `FocusModel:195` and then again in `FocusSubscriber:191`. Fixed in https://github.com/mautic/mautic/pull/11945/commits/3b0b015482eb3bdd7548923a2baa28a7b598e1d9

@RCheesley, @oltmanns-leuchtfeuer, @magdalenaleonow, thanks for the tests. I hope we have fixed all the problems so it is ready to test again.

escopecz commented on plugins/MauticFocusBundle/Controller/FocusController.php:
```diff
@@ -2,21 +2,41 @@
 {
+    private CacheProvider $cacheProvider;
+
+    /** @phpstan-ignore-next-line */
```
What was the problem here?

patrykgruszka replied:
Parent class requires `MauticFactory`, which is deprecated (I'm not using it). I need to override the constructor to inject the `CacheProvider`. Do you know better workaround than just ignore this constructor?

escopecz replied:
Ah, the deprecation warning. I was asking the same question in https://github.com/mautic/mautic/pull/12580#issuecomment-1647705911

In the future we can ignore just specific rule but it seems we'd have to upgrade PHPSTAN to support that. https://phpstan.org/user-guide/ignoring-errors So let's go with this for now.

escopecz commented on plugins/MauticFocusBundle/Controller/AjaxController.php:
```diff
@@ -40,4 +41,91 @@ public function generatePreviewAction(Request $request): JsonResponse
         return $this->sendJsonResponse($responseContent);
     }
+
+    public function getViewsCountAction(Request $request, CacheProvider $cacheProvider): JsonResponse
```
Could you please trigger these 2 new actions with a functional test? It seems they are not being run in PHPUNIT.

patrykgruszka replied:
Sure, done

escopecz replied:
You are fast! Thanks!

escopecz commented on plugins/MauticFocusBundle/Tests/Model/FocusModelFunctionalTest.php:
```diff
@@ -0,0 +1,131 @@
+{
+    public function testViewsCount(): void
+    {
+        /** @var FocusModel $focusModel */
+        $focusModel = self::$container->get('mautic.focus.model.focus');
```
As commented above, I'd change these tests to calling the new controller actions to test the whole feature like a user would instead of calling just the model method.

escopecz reviewed (changes requested):
Thanks for fixing those bugs!

patrykgruszka committed:
fix: [DPMMA-1742] change functional test

escopecz reviewed (approved):
Thank you! 👍

escopecz commented on plugins/MauticFocusBundle/Controller/FocusController.php:
```diff
@@ -134,22 +156,43 @@ public function getViewArguments(array $args, $action)
+                    $trackables = $trackableModel->getTrackableList('focus', $item->getId());
+                }
+
+                $cacheItem->set([$stats, $trackables]);
```
This introduced new regression:
```
500 Internal Server Error - PHP Warning - Undefined variable $trackables
plugins/MauticFocusBundle/Controller/FocusController.php:185 at Mautic\CoreBundle\ErrorHandler\ErrorHandler -> handleError ( 2, 'PHP Warning - Undefined variable $trackables', '/var/www/html/plugins/MauticFocusBundle/Controller/FocusController.php', 185 )
```

escopecz replied:
Fixed in https://github.com/mautic/mautic/pull/12769
